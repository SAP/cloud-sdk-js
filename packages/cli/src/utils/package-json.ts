/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { writeFileSync, readFileSync } from 'fs';
import { resolve } from 'path';
import execa from 'execa';
import { recordWarning } from '../utils';
import { getJestConfig } from './jest-config';

interface PackageJsonChange {
  scripts?: { [key: string]: string };
  dependencies?: string[];
  devDependencies?: string[];
  jest?: { [key: string]: any };
}

const frontendTestScripts: PackageJsonChange = {
  scripts: {
    'ci-frontend-unit-test':
      'echo "Test your application and write results in a JUnit format to `s4hana_pipeline/reports/frontend-unit/` and coverage in a cobertura format to `s4hana_pipeline/reports/coverage/frontend-unit/`"',
    'ci-e2e':
      'echo "Test your application and write results in a JUnit format to `s4hana_pipeline/reports/e2e/`"'
  }
};

const cdsChanges = {
  scripts: {
    'cds-build': 'cds build/all',
    'cds-deploy': 'cds deploy',
    'ci-build': 'npm run cds-deploy && npm run cds-build && npm run build',
    'ci-package':
      'sap-cloud-sdk package --include="package.json,package-lock.json,index.js,dist/**/*,db/**/*,srv/**/*,.cdsrc.json"'
  },
  devDependencies: ['sqlite3'],
  dependencies: ['@sap/cds', '@sap/cds-dk']
};

const scaffoldProjectPackageJson: PackageJsonChange = {
  scripts: {
    deploy: 'npm run build && sap-cloud-sdk package && cf push',
    'ci-build': 'npm run build',
    'ci-package': 'sap-cloud-sdk package --ci',
    'ci-it-backend': 'jest --ci --config ./test/jest-e2e.json',
    'ci-backend-unit-test': 'jest --ci'
  },
  devDependencies: [
    'jest-junit',
    '@sap-cloud-sdk/test-util',
    '@sap-cloud-sdk/cli'
  ],
  dependencies: ['@sap-cloud-sdk/core'],
  jest: getJestConfig(true)
};

const existingProjectPackageJson: PackageJsonChange = {
  scripts: {
    'ci-build': 'echo "Use this to compile or minify your application"',
    'ci-package':
      'sap-cloud-sdk package --include="package.json,package-lock.json,index.js,dist/**/*"',
    'ci-it-backend':
      'echo "Test your application and write results in a JUnit format to `s4hana_pipeline/reports/backend-integration/` and coverage in a cobertura format to `s4hana_pipeline/reports/coverage/backend-integration/`"',
    'ci-backend-unit-test':
      'echo "Test your application and write results in a JUnit format to `s4hana_pipeline/reports/backend-unit/` and coverage in a cobertura format to `s4hana_pipeline/reports/coverage/backend-unit/`"'
  },
  devDependencies: ['@sap-cloud-sdk/test-util', '@sap-cloud-sdk/cli'],
  dependencies: ['@sap-cloud-sdk/core']
};

export async function parsePackageJson(projectDir: string) {
  try {
    return JSON.parse(
      readFileSync(resolve(projectDir, 'package.json'), {
        encoding: 'utf8'
      })
    );
  } catch (error) {
    throw new Error(
      `Your package.json does not contain valid JSON. Please repair or delete it. ${error}`
    );
  }
}

function findScriptConflicts(originalScripts: any, scriptsToBeAdded: any) {
  return originalScripts
    ? Object.keys(scriptsToBeAdded).filter(name =>
        Object.keys(originalScripts).includes(name)
      )
    : [];
}

async function getPackageJsonChanges(
  isScaffold: boolean,
  frontendScripts: boolean,
  addCds: boolean
) {
  const changes: PackageJsonChange[] = [
    isScaffold ? scaffoldProjectPackageJson : existingProjectPackageJson
  ];
  if (frontendScripts) {
    changes.push(frontendTestScripts);
  }

  if (addCds) {
    changes.push(cdsChanges);
  }

  const merged = changes.reduce((mergedChanges, change) => {
    (Object.entries(change) as [keyof PackageJsonChange, any][]).forEach(
      ([key, value]) => {
        const newValue = { ...mergedChanges[key], ...value };
        mergedChanges[key] = Array.isArray(value)
          ? Object.values(newValue)
          : newValue;
      }
    );

    return mergedChanges;
  }, {});

  return {
    ...merged,
    dependencies: await addDependencyVersions(merged.dependencies),
    devDependencies: await addDependencyVersions(merged.devDependencies)
  };
}

function mergePackageJson(originalPackageJson: any, changes: any) {
  const adjustedPackageJson = {
    ...originalPackageJson,
    scripts: { ...originalPackageJson.scripts, ...changes.scripts },
    dependencies: {
      ...changes.dependencies,
      ...originalPackageJson.dependencies
    },
    devDependencies: {
      ...changes.devDependencies,
      ...originalPackageJson.devDependencies
    }
  };

  if (changes.jest) {
    adjustedPackageJson.jest = { ...originalPackageJson.jest, ...changes.jest };
  }

  return adjustedPackageJson;
}

export async function modifyPackageJson({
  projectDir,
  isScaffold = false,
  frontendScripts = false,
  force = false,
  addCds = false
}: {
  projectDir: string;
  isScaffold?: boolean;
  frontendScripts?: boolean;
  force?: boolean;
  addCds?: boolean;
}) {
  const originalPackageJson = await parsePackageJson(projectDir);
  const changes = await getPackageJsonChanges(
    isScaffold,
    frontendScripts,
    addCds
  );
  const conflicts = findScriptConflicts(
    originalPackageJson.scripts,
    changes.scripts
  );

  if (conflicts.length && !force) {
    throw new Error(
      conflicts.length > 1
        ? `Scripts with the names "${conflicts.join(
            '", "'
          )}" already exist. If you want to overwrite them, rerun the command with \`--force\`.`
        : `A script with the name "${conflicts.join(
            '", "'
          )}" already exists. If you want to overwrite it, rerun the command with \`--force\`.`
    );
  }

  writeFileSync(
    resolve(projectDir, 'package.json'),
    JSON.stringify(mergePackageJson(originalPackageJson, changes), null, 2)
  );
}

async function addDependencyVersions(
  dependencies: string[] = []
): Promise<{ [key: string]: string }> {
  const versions = await Promise.all(
    dependencies.map(dependency => getVersionOfDependency(dependency))
  );
  return dependencies.reduce(
    (result, dependency, index) => ({
      ...result,
      [dependency]: versions[index]
    }),
    {} as any
  );
}

async function getVersionOfDependency(dependency: string): Promise<string> {
  try {
    const args = ['view', dependency, 'version'];
    const version = dependency.includes('@sap/')
      ? execa('npm', [...args, '--registry', 'https://npm.sap.com'])
      : execa('npm', args);

    return `^${(await version).stdout}`;
  } catch (e) {
    recordWarning(
      `Could not find version information for dependency ${dependency}.`,
      'Instead `LATEST` was used as fallback.'
    );
    return 'latest';
  }
}

export async function installDependencies(
  projectDir: string,
  verbose: boolean
) {
  return execa('npm', ['install'], {
    cwd: projectDir,
    stdio: verbose ? 'inherit' : 'ignore'
  });
}
