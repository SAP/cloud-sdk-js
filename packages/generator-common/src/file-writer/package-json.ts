/**
 * @internal
 */
export interface PackageJsonOptions {
  /**
   * @internal
   */
  npmPackageName: string;
  /**
   * @internal
   */
  version: string;
  /**
   * @internal
   */
  sdkVersion: string;
  /**
   * @internal
   */
  description: string;
  /**
   * @internal
   */
  license?: string;
}

/* eslint-disable valid-jsdoc */
/**
 * @internal
 */
export function packageJsonBase(
  options: PackageJsonOptions
): Record<string, any> {
  return {
    name: options.npmPackageName,
    version: options.version,
    description: options.description,
    homepage: 'https://sap.github.io/cloud-sdk/docs/js/getting-started',
    main: './index.js',
    types: './index.d.ts',
    ...(options.license ? { license: options.license } : {}),
    publishConfig: {
      access: 'public'
    },
    repository: {
      type: 'git',
      url: ''
    }
  };
}
