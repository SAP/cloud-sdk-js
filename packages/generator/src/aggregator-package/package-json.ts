/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */

export function packageJson(
  npmPackageName: string,
  dependencies: string[],
  versionInPackageJson: string | undefined,
  generatorVersion: string
): string {
  return (
    JSON.stringify(
      {
        name: npmPackageName,
        version: versionInPackageJson || generatorVersion,
        description: 'SAP Cloud SDK for JavaScript: Complete Virtual Data Model (VDM)',
        homepage: 'https://www.sap.com/cloud-sdk',
        repository: {
          type: 'git',
          url: ''
        },
        scripts: {
          version: 'node ../../../after-version-update.js'
        },
        dependencies: dependencies.reduce((deps, service) => {
          deps[service] = `^${generatorVersion}`;
          return deps;
        }, {}),
        peerDependencies: {
          '@sap-cloud-sdk/core': `^${generatorVersion}`
        }
      },
      null,
      2
    ) + '\n'
  );
}
