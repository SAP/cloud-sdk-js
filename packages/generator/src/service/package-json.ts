import { ODataVersion, unixEOL } from '@sap-cloud-sdk/util';
import {
  packageJsonBase,
  PackageJsonOptions as PackageJsonOptionsBase
} from '@sap-cloud-sdk/generator-common/internal';

/**
 * @internal
 */
export interface PackageJsonOptions extends PackageJsonOptionsBase {
  /**
   * @internal
   */
  sdkAfterVersionScript: boolean;
  /**
   * @internal
   */
  oDataVersion: ODataVersion;
}

/**
 * Generate the package.json for an odata client so it can be released as an npm module.
 * @param options - Options to generate the package.json
 * @returns The package.json contents.
 * @internal
 */
export async function packageJson(
  options: PackageJsonOptions
): Promise<string> {
  const oDataModule =
    options.oDataVersion === 'v2'
      ? '@sap-cloud-sdk/odata-v2'
      : '@sap-cloud-sdk/odata-v4';
  return (
    JSON.stringify(
      {
        ...packageJsonBase(options),
        files: [
          '**/*.js',
          '**/*.js.map',
          '**/*.d.ts',
          '**/d.ts.map',
          '**/*-csn.json'
        ],
        scripts: {
          compile: 'npx tsc',
          ...(options.sdkAfterVersionScript
            ? { version: 'node ../../../after-version-update.js' }
            : {})
        },
        dependencies: {
          '@sap-cloud-sdk/odata-common': `^${options.sdkVersion}`,
          [oDataModule]: `^${options.sdkVersion}`
        },
        devDependencies: {
          typescript: '~4.5'
        }
      },
      null,
      2
    ) + unixEOL
  );
}
