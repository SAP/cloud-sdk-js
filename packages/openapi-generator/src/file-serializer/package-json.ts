import { unixEOL } from '@sap-cloud-sdk/util';
import {
  PackageJsonOptions,
  packageJsonBase
} from '@sap-cloud-sdk/generator-common/internal';

/**
 * Generate the package.json for an openapi client so it can be released as an npm module.
 * @param options - Options to generate the package.json
 * @returns The package.json contents.
 * @internal
 */
export function packageJson(options: PackageJsonOptions): string {
  return (
    JSON.stringify(
      {
        ...packageJsonBase(options),
        files: ['**/*.js', '**/*.js.map', '**/*.d.ts', '**/d.ts.map'],

        scripts: {
          compile: 'npx tsc'
        },
        dependencies: {
          '@sap-cloud-sdk/openapi': `^${options.sdkVersion}`
        },
        devDependencies: {
          typescript: '~4.1.2'
        }
      },
      null,
      2
    ) + unixEOL
  );
}
