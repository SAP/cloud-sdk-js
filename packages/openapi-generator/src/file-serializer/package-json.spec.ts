import { packageJson } from './package-json';

describe('packageJson', () => {
  it('returns the package.json content', () => {
    expect(
      packageJson({
        npmPackageName: 'workflow-service',
        description: 'description',
        sdkVersion: '1.35.0',
        version: '1.23.1'
      })
    ).toMatchSnapshot();
  });

  it('includes the license', () => {
    const parsed = JSON.parse(
      packageJson({
        npmPackageName: 'workflow-service',
        description: 'description',
        sdkVersion: '1.35.0',
        version: '1.23.1',
        license: 'my license information'
      })
    );
    expect(parsed.license).toBe('my license information');
  });
});
