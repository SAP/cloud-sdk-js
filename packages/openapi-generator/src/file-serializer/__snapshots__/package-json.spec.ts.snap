// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packageJson returns the package.json content 1`] = `
"{
  \\"name\\": \\"workflow-service\\",
  \\"version\\": \\"1.23.1\\",
  \\"description\\": \\"description\\",
  \\"homepage\\": \\"https://sap.github.io/cloud-sdk/docs/js/getting-started\\",
  \\"main\\": \\"./index.js\\",
  \\"types\\": \\"./index.d.ts\\",
  \\"publishConfig\\": {
    \\"access\\": \\"public\\"
  },
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"\\"
  },
  \\"files\\": [
    \\"**/*.js\\",
    \\"**/*.js.map\\",
    \\"**/*.d.ts\\",
    \\"**/d.ts.map\\"
  ],
  \\"scripts\\": {
    \\"compile\\": \\"npx tsc\\"
  },
  \\"dependencies\\": {
    \\"@sap-cloud-sdk/openapi\\": \\"^1.35.0\\"
  },
  \\"devDependencies\\": {
    \\"typescript\\": \\"~4.1.2\\"
  }
}
"
`;
